File: 105-construct_binary_tree_from_preorder_and_inorder_taversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# time: O(N^2)
# space O(N) - recursive stack for a highly unbalanced tree (not counting output)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# time: O(N^2)
# space O(N) - recursive stack for a highly unbalanced tree (not counting output)
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        if not inorder:  # Base case for stopping recursion
            return None

        # Find index of root node ('s value) within inorder
        # and remove node from inorder.
        # Front of preorder holds root's value, consume one node at a time
        inorder_idx = inorder.index(preorder.pop(0))
        root = TreeNode(inorder[inorder_idx])

        # Recursively generate left subtree starting from
        # 0th index to root index within inorder
        root.left = self.buildTree(preorder, inorder[:inorder_idx])

        # Recursively generate right subtree starting from
        # next of root index till last index
        root.right = self.buildTree(preorder, inorder[inorder_idx + 1:])

        return root

File: 105-construct_binary_tree_from_preorder_and_inorder_taversal.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {
    public TreeNode BuildTree(int[] preorder, int[] inorder)
    {
        
        if (inorder.Length == 0)
        {
            return null;
        }

        TreeNode node = new TreeNode(preorder[0]);
        preorder = preorder.Skip(1).ToArray();
        int inorderIdx = Array.IndexOf(inorder, node.val);

        node.left = BuildTree(preorder.Take(inorderIdx).ToArray(), inorder.Take(inorderIdx).ToArray());
        node.right = BuildTree(preorder.Skip(inorderIdx).ToArray(), inorder.Skip(inorderIdx + 1).ToArray());
        
        return node;
    }
}
